### ISSUE TITLE
Plugin doesn't work with Logstash 8.10+

### ISSUE THREAD

**Issue by mashhurs** [logstash-plugins/logstash-input-imap#61]
### Description
**TL;DR: Ruby 3.1+ compatible `mail` 2.8 breaks the plugin behavior.**

Details: Logstash 8.10+ versions use Ruby 3.1. Ruby 3.1 extracts bundled gems such as `net-smtp`, `net-imap` where `mail` (this plugin depends on) needs to explicitly include them. The `mail` gem includes Ruby 3.1 support in 2.8+, however the big issue, when this plugin upgrades `mail` to 2.8+ ([upgrade PR](https://github.com/logstash-plugins/logstash-input-imap/pull/60)), its behaviour is completely broken. The [test cases](https://app.travis-ci.com/github/logstash-plugins/logstash-input-imap/jobs/610779677) show,
- email content type is null eligible
- attachments are automatically decoded
- need more investigation

###

**Comment by ceeeekay** [comment]
Is there any progress here? One of my Logstash nodes is still pinned on 8.9.2, as the data it collects is only available via an IMAP account.

**Comment by yaauie** [comment]
Moving forward with confidence with this one is going to be very difficult, primarily because the specs of this plugin use `Mail.new(&block)` to _dynamically compose_ the mail `Message` objects that we test against, and newer versions of `mail` compose different raw messages than older versions. Because this plugin is used to _parse_ messages, we need to ensure it behaves the same when encountering the exact same sequence of bytes.

Therefore, we would need to do the following _in order_:

 1. with an old logstash, capture the on-the-wire form of the messages in specs as _fixtures_ and rework the spec _setup_ to load each message from its fixture _without modifying any of the specs themselves_.
 2. come up with a patch for the implementation that allows the specs-from-fixtures to pass with `mail` ~> `2.8` without modifying the specs or the fixtures

**Comment by ceeeekay** [comment]
Is anything at all happening here? It's been well over a year now and this plugin is completely unusable.

I still have a production node pinned on LS 8.9 which is starting to raise some eyebrows in our security team.

**Comment by anrxc** [comment]
Our compliance team is very not happy with us either. If feature request +1 counts for anything, there you go.

**Comment by ceeeekay** [comment]
Another 6 months with no updates and no resolution in sight. Over a year since a dev last commented on this issue.

What's happening here? Is it too hard? Have you completely given up?

This is quite poor.

**Comment by rado-van** [comment]
seems like this won't be supported anymore, last activity was 2 years ago.. we had to come up with a custom solution

**Comment by idahomst** [comment]
Hi everyone,

## tl;dr:
```ruby
 ../../../../usr/share/logstash/vendor/bundle/jruby/3.1.0/gems/net-imap-0.5.13/lib/net/imap.rb                                               
╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌
 3787          if defined?(VerifyCallbackProc)                                                                                               
 3788            context.verify_callback = VerifyCallbackProc                                                                                
 3789          end                                                                                                                           
 3790 +        context.setup if context.respond_to?(:setup) # JRuby: must call before freeze                                                 
 3791          context.freeze                                                                                                                
 3792          [params, context]                                                                                                             
 3793        else                                                                                                                            
╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌
```

## Long story:
This bug is making me a bit angry. As a Linux admin, I’m not able to fully understand the library dependency chain, so I set up a test environment and spent some tokens with Claude Code. As a result, I now have a working IMAP connection to my mail server and I created these two issues: 
* https://github.com/ruby/net-imap/issues/624
* https://github.com/jruby/jruby-openssl/issues/345


---


## Full bug description from Claude Code:
# Logstash IMAP Input — SSL NullPointerException: Root Cause Analysis & Fix

## Environment

| Component | Version |
|---|---|
| Logstash | 9.3.1 |
| logstash-input-imap | 3.2.1 |
| net-imap gem | 0.5.13 |
| jruby-openssl gem | 0.15.5 |
| JRuby | 9.4.13.0 |
| Java (OpenJDK) | 21.0.10 |
| OS | Debian Linux (kernel 6.12) |

---

## The Error

On every IMAP poll cycle, Logstash threw:

```
[ERROR][logstash.inputs.imap] Encountered error Java::JavaLang::NullPointerException
  {:message=>"Cannot invoke
    \"org.jruby.ext.openssl.SSLContext$InternalContext.getSSLContext()\"
    because \"this.internalContext\" is null",
   :backtrace=>[
    "org.jruby.ext.openssl.SSLContext.createSSLEngine(SSLContext.java:784)",
    "org.jruby.ext.openssl.SSLSocket.ossl_ssl_setup(SSLSocket.java:221)",
    "org.jruby.ext.openssl.SSLSocket.connectImpl(SSLSocket.java:299)",
    "org.jruby.ext.openssl.SSLSocket.connect_nonblock(SSLSocket.java:288)",
    ...
    "org.jruby.ir.instructions.UnresolvedSuperInstr.interpret(...)",  # Ruby super call
    "org.jruby.RubyClass.newInstance(RubyClass.java:936)",
    ...
  ]}
```

The plugin configuration (`/etc/logstash/conf.d/pipeline.conf`):

```ruby
input {
    imap {
        host        => "imap.server"
        user        => "user@example.com"
        password    => "..."
        secure      => true        # use IMAPS, port 993
        verify_cert => false       # skip certificate verification
    }
}
```

---

## Dependency Chain

The bug lives at the intersection of four components. Understanding which library
is actually loaded at runtime is the key to finding it.

```
logstash-input-imap 3.2.1
│
├── requires 'net/imap'          ← expects Ruby stdlib
│   │
│   └── BUT Logstash's bundler environment has on the load path:
│       net-imap gem 0.5.13      ← this wins, loaded instead of stdlib
│           │
│           └── pulled in transitively by:
│               mail gem 2.9.0   ← runtime dependency of logstash-input-imap
│
├── requires 'mail'              ← direct dependency
│
└── uses jruby-openssl 0.15.5   ← JRuby's OpenSSL implementation
        │
        └── wraps BouncyCastle (bcprov-jdk18on 1.81)
```

At runtime, `Net::IMAP` resolves to `net-imap 0.5.13`, not the Ruby stdlib version.
You can verify this with:

```ruby
require 'net/imap'
puts Net::IMAP.instance_method(:initialize).source_location[0]
# => ".../gems/net-imap-0.5.13/lib/net/imap/deprecated_client_options.rb"
```

---

## How the Connection Is Built

### Plugin's `connect` method (`imap.rb:123`)

```ruby
def connect
  sslopt = @secure                              # sslopt = true
  if @secure and not @verify_cert
    sslopt = { :verify_mode => OpenSSL::SSL::VERIFY_NONE }   # sslopt = hash
  end
  imap = Net::IMAP.new(@host, :port => @port, :ssl => sslopt)
  imap.login(@user, @password.value)
  return imap
end
```

With `secure: true` and `verify_cert: false`, the call becomes:

```ruby
Net::IMAP.new(host, port: 993, ssl: { verify_mode: OpenSSL::SSL::VERIFY_NONE })
```

### What `net-imap 0.5.13` does with that hash

```ruby
# net-imap-0.5.13/lib/net/imap.rb

def initialize(host, port: nil, ssl: nil, ...)
  ...
  @ssl_ctx_params, @ssl_ctx = build_ssl_ctx(ssl)   # ssl = { verify_mode: 0 }
  ...
  start_tls_session if ssl_ctx
end

def build_ssl_ctx(ssl)
  if ssl
    params  = (Hash.try_convert(ssl) || {}).freeze
    context = SSLContext.new
    context.set_params(params)                 # sets verify_mode = VERIFY_NONE
    if defined?(VerifyCallbackProc)
      context.verify_callback = VerifyCallbackProc
    end
    context.freeze                             # <<< THE BUG IS HERE
    [params, context]
  else
    false
  end
end

def start_tls_session
  ...
  @sock = SSLSocket.new(@sock, ssl_ctx)        # passes the frozen context
  ssl_socket_connect(@sock, open_timeout)      # calls connect_nonblock in a loop
end
```

### The `ssl_socket_connect` loop (stdlib `net/protocol.rb`)

```ruby
def ssl_socket_connect(s, timeout)
  while true
    raise Net::OpenTimeout if timeout <= 0
    start = Process.clock_gettime Process::CLOCK_MONOTONIC
    case s.connect_nonblock(exception: false)   # <<< crashes here
    when :wait_readable; s.to_io.wait_readable(timeout)
    when :wait_writable; s.to_io.wait_writable(timeout)
    else; break
    end
    timeout -= Process.clock_gettime(Process::CLOCK_MONOTONIC) - start
  end
end
```

---

## Root Cause: `freeze` vs. JRuby Lazy Initialization

### How JRuby OpenSSL initializes an SSL connection

In JRuby's `SSLContext` Java implementation, the internal BouncyCastle SSL context
(`internalContext`) is **lazily initialized** — it is created the first time it is
actually needed, which is during `createSSLEngine()`, called from `connect_nonblock`.

```
SSLSocket.connect_nonblock(exception: false)
  └── SSLSocket.connectImpl()
        └── SSLSocket.ossl_ssl_setup()
              └── SSLContext.createSSLEngine()    # line 784
                    └── this.internalContext.getSSLContext()
                                ^
                                NullPointerException here
                                because internalContext was never set
```

### Why `freeze` prevents initialization

Ruby's `Object#freeze` makes an object fully immutable. JRuby enforces this at the
VM level: **any attempt to write an instance variable on a frozen object raises
`FrozenError` or is silently blocked**.

`net-imap 0.5.13` freezes the `SSLContext` before the connection is made:

```
Timeline:

  build_ssl_ctx()
    SSLContext.new          → internalContext = nil   (not yet set)
    context.set_params(..)  → sets Ruby-level attributes
    context.freeze          → object is now immutable ← PROBLEM
    return context

  start_tls_session()
    SSLSocket.new(sock, context)
    connect_nonblock        → tries to call createSSLEngine()
                               which needs to set internalContext
                               but object is frozen → stays nil
                               → NullPointerException
```

### Proof of concept

```ruby
require 'openssl'
require 'socket'

HOST = 'imap.server'
PORT = 993

# FAILS — frozen context, internalContext never initialized
tcp = TCPSocket.new(HOST, PORT)
ctx = OpenSSL::SSL::SSLContext.new
ctx.set_params(verify_mode: OpenSSL::SSL::VERIFY_NONE)
ctx.freeze                                   # net-imap does this
ssl = OpenSSL::SSL::SSLSocket.new(tcp, ctx)
ssl.connect_nonblock(exception: false)
# => NullPointerException: Cannot invoke ...getSSLContext()... because
#    "this.internalContext" is null

# WORKS — no freeze
tcp = TCPSocket.new(HOST, PORT)
ctx = OpenSSL::SSL::SSLContext.new
ctx.set_params(verify_mode: OpenSSL::SSL::VERIFY_NONE)
# ctx.freeze                                 # omitted
ssl = OpenSSL::SSL::SSLSocket.new(tcp, ctx)
ssl.connect_nonblock(exception: false)
# => :wait_readable  (normal, handshake in progress)
```

---

## The Fix

`SSLContext#setup` forces eager initialization of the internal SSL engine,
so that when `freeze` is called immediately after, `internalContext` is already set
and no further mutation is needed during `connect_nonblock`.

**File:** `net-imap-0.5.13/lib/net/imap.rb` — method `build_ssl_ctx`

```ruby
# BEFORE
def build_ssl_ctx(ssl)
  if ssl
    params  = (Hash.try_convert(ssl) || {}).freeze
    context = SSLContext.new
    context.set_params(params)
    if defined?(VerifyCallbackProc)
      context.verify_callback = VerifyCallbackProc
    end
    context.freeze
    [params, context]
  else
    false
  end
end

# AFTER
def build_ssl_ctx(ssl)
  if ssl
    params  = (Hash.try_convert(ssl) || {}).freeze
    context = SSLContext.new
    context.set_params(params)
    if defined?(VerifyCallbackProc)
      context.verify_callback = VerifyCallbackProc
    end
    context.setup if context.respond_to?(:setup)  # JRuby: init internalContext before freeze
    context.freeze
    [params, context]
  else
    false
  end
end
```

The `respond_to?(:setup)` guard ensures the fix is safe on MRI Ruby (CRuby), where
`SSLContext#setup` exists but has slightly different semantics, and on any future
version that may not expose the method at all.

### Verification

```ruby
# Simulate Logstash's bundler environment
$LOAD_PATH.unshift('/usr/share/logstash/vendor/bundle/jruby/3.1.0/gems/net-imap-0.5.13/lib')
require 'net/imap'
require 'openssl'

sslopt = { :verify_mode => OpenSSL::SSL::VERIFY_NONE }
imap = Net::IMAP.new('mailmx1-tt2.tt2.lssrv.tech', :port => 993, :ssl => sslopt)
# => connected
imap.login('user@example.com', '...')
# => login OK
imap.select('INBOX')
# => INBOX selected, N messages
imap.logout; imap.disconnect
```

---

## Important: Fix Persistence

The fix was applied directly to the installed gem file:

```
/usr/share/logstash/vendor/bundle/jruby/3.1.0/gems/
    net-imap-0.5.13/lib/net/imap.rb
```

**This fix will be lost if:**
- `logstash-plugin update logstash-input-imap` is run (reinstalls `mail` → `net-imap`)
- `logstash-plugin install` for any plugin triggers a full bundle update
- Logstash itself is upgraded

**Recommended long-term actions:**

1. **Report upstream** — file a bug against `net-imap` gem
   (`https://github.com/ruby/net-imap`) describing that `context.freeze` before
   `connect_nonblock` causes a `NullPointerException` under JRuby due to lazy
   initialization of `internalContext` in `jruby-openssl`.

2. **Report upstream** — alternatively (or additionally) file against `jruby-openssl`
   (`https://github.com/jruby/jruby-openssl`) that freezing an `SSLContext` after
   `set_params` but before first use should not cause an NPE — `internalContext`
   should be initialized eagerly in `set_params` or in the constructor.

3. **Local safeguard** — add a post-install script or Puppet/Ansible task that
   re-applies the one-line patch after any Logstash plugin update:

```bash
#!/bin/bash
# Re-apply net-imap freeze fix for JRuby compatibility
FILE="/usr/share/logstash/vendor/bundle/jruby/3.1.0/gems/net-imap-0.5.13/lib/net/imap.rb"
if grep -q "context.setup if context.respond_to" "$FILE"; then
    echo "Patch already applied."
else
    sed -i 's/        context\.freeze/        context.setup if context.respond_to?(:setup) # JRuby: init internalContext before freeze\n        context.freeze/' "$FILE"
    echo "Patch applied."
fi
```

---

Hope that helps.

**Comment by ceeeekay** [comment]
For anyone following along, this plugin is working again since Logstash 9.4.0 because the underlying net-imap gem has been upgraded to 0.6.4.1 for JRuby 10.

I don't know if that guarantees it will continue to work in future versions, so I've pinned mine to 9.4.3. At least that will get our security team off my back for a while.
